src/modules/gradle.rs

```diff
@@ -0,0 +1,212 @@
+    module::Module,
+};
+
+pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
+    let mut module = context.new_module("gradle");
+    let config = GradleConfig::try_load(module.config);
+    let is_gradle_project = context
+        .try_begin_scan()?
+        .set_files(&config.detect_files)
+        .set_extensions(&config.detect_extensions)
+        .set_folders(&config.detect_folders)
+        .is_match();
+
+    if !is_gradle_project {
+        return None;
+    }
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        formatter
+            .map_meta(|var, _| match var {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "version" => {
+                    let gradle_version = {
+                        let properties_file =
+                            find_wrapper_properties_file(&context.current_dir, config.recursive)?;
+                        let properties = fs::read_to_string(properties_file).ok()?;
+                        parse_gradle_version_from_properties(&properties)?
+                    };
+                    VersionFormatter::format_module_version(
+                        module.get_name(),
+                        &gradle_version,
+                        config.version_format,
+                    )
+                    .map(Ok)
+                }
+                _ => None,
+            })
+            .parse(None, Some(context))
+    });
+
+    module.set_segments(match parsed {
+        Ok(segments) => segments,
+        Err(error) => {
+            log::warn!("Error in module `gradle`:\n{}", error);
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+fn parse_gradle_version_from_properties(wrapper_properties: &str) -> Option<String> {
+    // example gradle.properties content
+    /*
+        distributionBase=GRADLE_USER_HOME
+        distributionPath=wrapper/dists
+        distributionUrl=https\://services.gradle.org/distributions/gradle-7.5.1-bin.zip
+        zipStoreBase=GRADLE_USER_HOME
+        zipStorePath=wrapper/dists
+    */
+    let version = wrapper_properties
+        .split_once("distributionUrl=")?
+        .1
+        .split_once("gradle-")?
+        .1
+        .split_once('-')?
+        .0;
+    Some(version.to_string())
+}
+
+/// Tries to find the gradle-wrapper.properties file.
+fn find_wrapper_properties_file(directory: &Path, recursive: bool) -> Option<PathBuf> {
+    let properties_path = directory.join("gradle/wrapper/gradle-wrapper.properties");
+    if properties_path.exists() {
+        return Some(properties_path);
+    }
+
+    if recursive {
+        find_wrapper_properties_file(directory.parent()?, recursive)
+    } else {
+        None
+    }
+}
```

Comment:
I would prefer this sort of thing, making use of the scanned directory contents and reading the file straight away:
```suggestion
fn get_wrapper_properties_file(context: &Context, recursive: bool) -> Option<String> {
    let mut properties = None;
    if context
        .try_begin_scan()?
        .set_folders(&["gradle"])
        .is_match()
    {
        properties =
            utils::read_file(&context.current_dir.join("gradle/wrapper/gradle-wrapper.properties")).ok();
    };

    if recursive && properties.is_none() {
        for dir in context.current_dir.ancestors().skip(1) {
            properties =
                utils::read_file(&dir.join("gradle/wrapper/gradle-wrapper.properties")).ok();
            if properties.is_some() {
                break;
            }
        }
    }

    properties
}
```